component/discovery/discovery.go

```diff
@@ -0,0 +1,67 @@
+package discovery
+
+import (
+	"context"
+	"time"
+
+	"github.com/grafana/agent/component/metrics/scrape"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+// RunDiscovery is a utility for consuming and forwarding target groups from a discoverer.
+// It will handle collating targets (and clearing), as well as time based throttling of updates.
+// The channel provided should be the same channel passed to the Discovery's Run method.
+func RunDiscovery(ctx context.Context, ch <-chan []*targetgroup.Group, f func([]scrape.Target)) {
```

Comment:
I kinda like the idea presented here; I _think_ it will easily allow us to plug in new SD mechanisms. Do we have anything else in mind that would enable portability?

Couple of small things; 
a) do we want to make the frequency configurable? I think that was your original intent here, right? 
b) I think we could improve readability by simplifying some `if dirty` and `nextValidTime` bits with a simple time.Ticker. Is it just so that we avoid an initial empty change log, or is there another reason?
c) Have we encountered cases where throttling is going to be required, or is it an early optimization?

Reply:
With the way I've done it now, infrequent updates would get sent right away, while frequent ones will be sent as soon as the time has elapsed. With a ticker, all updates would be stuck waiting for the next tick. Maybe that is a premature optimization though. It certainly makes the code quite a bit more cumbersome. 

I think the throttling is modeled from prometheus, and their "contract" that says discovery should have a maximum update frequency. @rfratto may know more. If that is the case, we may or may not want frequency to be too configurable depending on the rigidity of that requirement.

Reply:
We can consider making it configurable in the future, but not required for this PR IMO. It's also not currently configurable in the Prometheus SD manager either; no more than once every 5 sec is [hard coded](https://github.com/prometheus/prometheus/blob/00ec720c29bce637c1580a2ce43b32122d133f83/discovery/manager.go#L119).

Reply:
I would personally prefer a snippet like the one presented below; I don't think it would be too much of a performance hit.